src/jit/namedintrinsiclist.h

```diff
@@ -66,6 +66,20 @@ enum HWIntrinsicFlag : unsigned int
     // Copy Upper bits
     // some SIMD scalar intrinsics need the semantics of copying upper bits from the source operand
     HW_Flag_CopyUpperBits = 0x200,
+
+    // Invert integral operands
+    HW_Flag_InvertIntegralOperands = 0x400,
+
+    // Select instruction using argument type
+    HW_Flag_InstrByArgType = 0x800,
+
+    // Flags for types supported by fixed immediate value instructions
+
+    // Float instructions
+    HW_Flag_Imm_Float = 0x010000,
+
+    // Double instructions
+    HW_Flag_Imm_Double = 0x020000,
```

Comment:
I do not think you need `HW_Flag_Imm_Float ` or `HW_Flag_Imm_Double`.

Reply:
I'm not sure that it is enough to handle all instructions with fixed immediate values via table correctly. We would need to comb the whole ISA range we want to support and find overlapping intrinsics, they signatures, corresponding instructions and compare it to they operands. Based on this we could design error proof method to handle diverse instructions which just happen to be in the same row of the table. Right now we have almost implemented 2 medium sized ISAs and there are at least 2 table problems which can be simply solved by additional information available in table.

The perfect solution could be based on expanded table and HWIntrinsicInfo structure. Currently missing information is to which instructions (or types) in `HWIntrinsicInfo.instrcutions` array fixed immediate values should be applied (see code below), which instructions require reversing operands order and which instructions should be selected using argument type rather than return type.

If we look at definition below it is obvious that we have no information in the table how to use `ival` or to which of the 10 instructions  should we apply it. To get this information we can check (i) what is instruction format - available in later phases in the compiler, (ii) gain this knowledge at this stage by querying for instruction format, (iii) add information to table to use it during importation and codegen.

If we go with the option (iii) what fits nicely into idea of table driven import and codegen than we have some choices. First we can go for perfect solution where `ival` would be an array of 10 values each corresponding to type encoded in instruction position. This perfect solution which should allow to handle any combination of `ival` and instruction type however size of `HWIntrinsicInfo` would increase by 40 bytes.

Another but less flexible solution is to add 16 bit bitfield after `ival` which could encode up to 16 positions for which to apply common fixed `ival`. But since we have mostly unused `HWIntrinsicFlag` enum it is much easier to reuse this field to encode types on demand.

This same reasoning can be applied to two remaining new flags `HW_Flag_ReverseIntegralOperandsOrder` and `HW_Flag_BaseTypeFromArg`.

The additional benefit coming with some of the flags and in particular with `HW_Flag_Imm_Float` or `HW_Flag_Imm_Double` is a type check which prevents applying immediate fixed operand to instruction not supporting given type.

```C++
struct HWIntrinsicInfo
{
    NamedIntrinsic      intrinsicID;
    const char*         intrinsicName;
    InstructionSet      isa;
    int                 ival;
    unsigned            simdSize;
    int                 numArgs;
    instruction         ins[10];
    HWIntrinsicCategory category;
    HWIntrinsicFlag     flags;
};
```

Reply:
The instructions which potentially take a fixed ival are already small. I'm not even sure having the ival in the table is worth it (this is a reversal from my earlier thoughts), given the small % of intrinsics it actually benefits.

It is probably easiest to just handle them specially... (@fiigii, thoughts?)

Reply:
> It is probably easiest to just handle them specially... (@fiigii, thoughts?)

Agree. But let me investigate a bit more.

Reply:
Fixed ival have a range of 0 .. 7. It means they can fit in 3 bits of  bitfield, if we combine it with one additional bit we can encode all values for floats (double and float) - just in 4 bits. Perhaps we can use top 4 bits of `HW_Flags` to encode all fixed immediate values needed by both float and double.

Reply:
It would provide support for at least `cmpps`, `cmpss`, `cmppd` and `cmpsd`.

Reply:
Or perhaps, since ival has 32 bits it could be used to encode any immediate value in the range to 16 bits, with remaining 16 bits left for type bitfield.

Reply:
> Fixed ival have a range of 0 .. 7.

We need more fixed imm values for SSE4.1 `Round*`.

Reply:
@fiigii You are right we need a lot of them even for the above four with VEX or EVEX encoding: from Intel manual

> The comparison predicate operand is an 8-bit immediate:
• For instructions encoded using the VEX or EVEX prefix, bits 4:0 define the type of comparison to be performed
(see Table 3-1). Bits 5 through 7 of the immediate are reserved.
• For instruction encodings that do not use VEX prefix, bits 2:0 define the type of comparison to be made (see the
first 8 rows of Table 3-1). Bits 3 through 7 of the immediate are reserved.

Reply:
> ival has 32 bits it could be used to encode any immediate value in the range to 16 bits, with remaining 16 bits left for type bitfield

But this encoding can handle both `type information` for up to 16 types (some reserve for new binary floating types left) and larger immediate operand range up to 16 bits.

Reply:
Only in the Sse we have 24 intrinsics entries in the table with fixed immediate values out of 81 entries what is roughly 30% of total.  For Sse2 this would be most probably 24 table entries as well out of roughly 100 total. 

My estimate would be that we should expect at least 120  table entries for all ISA.